ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/blobs/BlobsSidecarManagerImpl.java

```diff
@@ -0,0 +1,148 @@
+ */
+
+package tech.pegasys.teku.statetransition.blobs;
+
+import static java.util.Collections.emptyMap;
+import static tech.pegasys.teku.spec.logic.versions.eip4844.blobs.BlobsSidecarAvailabilityChecker.ALREADY_CHECKED;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Collections;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentSkipListMap;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.ethereum.events.SlotEventsChannel;
+import tech.pegasys.teku.infrastructure.collections.LimitedMap;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.datastructures.blocks.SlotAndBlockRoot;
+import tech.pegasys.teku.spec.datastructures.blocks.blockbody.versions.eip4844.BeaconBlockBodyEip4844;
+import tech.pegasys.teku.spec.datastructures.execution.versions.eip4844.BlobsSidecar;
+import tech.pegasys.teku.spec.logic.versions.eip4844.blobs.BlobsSidecarAvailabilityChecker;
+import tech.pegasys.teku.statetransition.forkchoice.ForkChoiceBlobsSidecarAvailabilityChecker;
+import tech.pegasys.teku.storage.api.StorageQueryChannel;
+import tech.pegasys.teku.storage.api.StorageUpdateChannel;
+import tech.pegasys.teku.storage.client.RecentChainData;
+
+public class BlobsSidecarManagerImpl implements BlobsSidecarManager, SlotEventsChannel {
+  private static final int MAX_CACHED_VALIDATED_BLOBS_PER_SLOT = 10;
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Spec spec;
+  private final RecentChainData recentChainData;
+  private final StorageQueryChannel storageQueryChannel;
+  private final StorageUpdateChannel storageUpdateChannel;
+
+  private final NavigableMap<UInt64, Map<Bytes32, BlobsSidecar>> validatedPendingBlobs =
+      new ConcurrentSkipListMap<>();
+
+  public BlobsSidecarManagerImpl(
+      final Spec spec,
+      final RecentChainData recentChainData,
+      final StorageQueryChannel storageQueryChannel,
+      final StorageUpdateChannel storageUpdateChannel) {
+
+    this.spec = spec;
+    this.recentChainData = recentChainData;
+    this.storageUpdateChannel = storageUpdateChannel;
+    this.storageQueryChannel = storageQueryChannel;
+  }
+
+  @Override
+  public void storeUnconfirmedValidatedBlobs(final BlobsSidecar blobsSidecar) {
+    // cache already validated blobs
+    validatedPendingBlobs
+        .computeIfAbsent(blobsSidecar.getBeaconBlockSlot(), __ -> createNewMap())
+        .put(blobsSidecar.getBeaconBlockRoot(), blobsSidecar);
+
+    internalStoreUnconfirmedBlobs(blobsSidecar);
+  }
+
+  @Override
+  public void storeUnconfirmedBlobs(final BlobsSidecar blobsSidecar) {
+    // remove blobs from the already validated cache
```

Comment:
Why? We couldn't unvalidate blobs, but we could get, say, IGNORE result on the second import try and remove validation in this case. I don't understand why it's needed. From my point of view, it could only damage us in some case

Reply:
yeah it isn't required.. it is a strange case, that should never occur, so the idea was to be conservative and force a revalidation of the sidecar against commitments during subsequent block import. But since it is strictly coupled with the block via blockroot I can remove it.